lib/Parse/ParsePattern.cpp

```diff
@@ -786,15 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
   auto result = parsePattern();
   
   // Now parse an optional type annotation.
-  if (consumeIf(tok::colon)) {
+  if (Tok.is(tok::colon)) {
+    SourceLoc colonLoc = consumeToken(tok::colon);
+    
     if (result.isNull())  // Recover by creating AnyPattern.
       result = makeParserErrorResult(new (Context) AnyPattern(PreviousLoc));
     
     ParserResult<TypeRepr> Ty = parseType();
     if (Ty.hasCodeCompletion())
       return makeParserCodeCompletionResult<Pattern>();
-    if (Ty.isNull())
+    if (!Ty.isNull()) {
+      // Attempt to diagnose initializer calls incorrectly written
+      // as typed patterns, such as "var x: [Int]()".
+      if (Tok.isFollowingLParen()) {
+        SourceLoc lParenLoc, rParenLoc;
+        SmallVector<Expr *, 2> args;
+        SmallVector<Identifier, 2> argLabels;
+        SmallVector<SourceLoc, 2> argLabelLocs;
+        Expr *trailingClosure;
+        ParserStatus status = parseExprList(tok::l_paren, tok::r_paren,
+                                            /*isPostfix=*/true,
+                                            /*isExprBasic=*/false,
+                                            lParenLoc, args, argLabels,
+                                            argLabelLocs, rParenLoc,
+                                            trailingClosure);
+        if (status.isSuccess()) {
+          // Suggest replacing : with = (with proper whitespace).
+          llvm::StringRef replacement = " =";
```

Comment:
@slavapestov Which Lexer methods do you mean? Isn't the lexer mostly unaware of whitespace?

I modeled this off https://github.com/apple/swift/blob/d3d2b0afd7e33e14f46c8b772fae965bdf38aa33/lib/Sema/CSDiag.cpp#L5616 and https://github.com/apple/swift/blob/d3d2b0afd7e33e14f46c8b772fae965bdf38aa33/lib/AST/DiagnosticEngine.cpp#L150